File: kotlin/matching-brackets/src/main/kotlin/MatchingBrackets.kt
object MatchingBrackets {

    fun isValid(input: String): Boolean {
        val stack = ArrayDeque<Char>()

        val pairs = mapOf(
            ')' to '(',
            '}' to '{',
            ']' to '[',
        )

        for (ch in input) {
            when (ch) {
                in pairs.values -> stack.addLast(ch)
                in pairs.keys -> {
                    if (stack.isEmpty() || stack.removeLast() != pairs[ch]) {
                        return false
                    }
                }
            }
        }

        return stack.isEmpty()
    }
}

File: csharp/matching-brackets/MatchingBrackets.cs
public static class MatchingBrackets
{
    public static bool IsPaired(string input)
    {
        Stack<char> stack = [];

        var pairs = new Dictionary<char, char>
        {
            { ')', '(' },
            { '}', '{' },
            { ']', '[' }
        };

        foreach (var ch in input)
        {
            if (pairs.ContainsValue(ch))
            {
                stack.Push(ch);
            }
            else if (pairs.ContainsKey(ch))
            {
                if (stack.Count == 0 || stack.Pop() != pairs[ch])
                {
                    return false;
                }
            }
        }

        return stack.Count == 0;
    }
}
